### PR TITLE
feat: stabilize global frontend api cache

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

* renamed to fronted api cache
* added comprehensive test suite
* added state transitions to make it testable and easier to use for production code

### CODE REVIEW

File: src/lib/proxy/global-frontend-api-cache.ts [hunk 1]
```diff
@@ -9,38 +9,36 @@ import {
...
-    private running: boolean;
+    private status: GlobalFrontendApiCacheState = 'starting';
 
     constructor(
         config: Config,
         segmentReadModel: ISegmentReadModel,
-        clientFeatureToggleReadModel: ClientFeatureToggleReadModel,
-        configurationRevisionService: ConfigurationRevisionService,
+        clientFeatureToggleReadModel: IClientFeatureToggleReadModel,
+        configurationRevisionService: EventEmitter,
```
Comment (kwasniew) [inline src/lib/proxy/global-frontend-api-cache.ts:41]: relaxed the contraints on the dependency. It's enough to make it event emitter

File: src/lib/proxy/client-feature-toggle-read-model-type.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { IFeatureToggleClient } from '../types';
+
+export interface IClientFeatureToggleReadModel {
```
Comment (kwasniew) [inline src/lib/proxy/client-feature-toggle-read-model-type.ts:3]: added missing interface

File: src/lib/proxy/frontend-api-repository.ts [hunk 3]
```diff
@@ -23,13 +23,13 @@ export class FrontendApiRepository
 
     private readonly token: IApiUser;
 
-    private globalFrontendApiRepository: GlobalFrontendApiRepository;
+    private globalFrontendApiRepository: GlobalFrontendApiCache;
```
Comment (kwasniew) [inline src/lib/proxy/frontend-api-repository.ts:26]: rename repository to cache

File: src/lib/proxy/global-frontend-api-cache.ts [hunk 4]
```diff
@@ -88,6 +90,13 @@ export class GlobalFrontendApiRepository extends EventEmitter {
         try {
             this.featuresByEnvironment = await this.getAllFeatures();
             this.segments = await this.getAllSegments();
+            if (this.status === 'starting') {
```
Comment (kwasniew) [inline src/lib/proxy/global-frontend-api-cache.ts:93]: cache is event emitter itself and events make it easy to write tests without artificial wait times

**Review by sjaanus — APPROVED** [review]

**Pull request merged**